### PR TITLE
Jittable eigh

### PR DESCRIPTION
**Pull request by adam-lewis-gcorp**
@hawkinsp @tlu7 @rmlarson

This commit evades the varying blocksize issue associated with QDWH-SVD by using a different polynomial iteration that computes a rank-k orthogonal projector for given k directly. The specific algorithm is described here: https://core.ac.uk/download/pdf/79508571.pdf.

I'd been trying to avoid this because the performance and error behaviour, while still not awful, tends not to be quite so good as obtained by QDWH. But the solutions I came up with turned out to be generally even worse.

### CODE REVIEW

File: jax/_src/scipy/eigh.py [hunk 1]
```diff
@@ -21,6 +21,32 @@
 from jax import lax
 
 
+def _fill_diagonal(X, vals):
+  return jax.ops.index_update(X, jnp.diag_indices(X.shape[0]), vals)
```
Comment (hawkinsp) [inline jax/_src/scipy/eigh.py:25]: JAX style thing: I'd always prefer `X.at[index].update(y)` over `index_update` these days. They are synonyms but I think the terser syntax is more readable.

File: tests/lax_scipy_test.py [hunk 2]
```diff
@@ -559,8 +554,13 @@ def test_spectral_dac_eigh(self, linear_size, seed, dtype):
       self.assertRaises(
         NotImplementedError, jax._src.scipy.eigh.eigh, H)
       return
-    evs, V = jax._src.scipy.eigh.eigh(H)
+    evs, V = jax._src.scipy.eigh.eigh(H, termination_size=2)
     ev_exp, eV_exp = jnp.linalg.eigh(H)
+    #guess = jax._src.scipy.eigh._canonically_purify_initial_guess(H, linear_size // 2)
```
Comment (rmlarsen) [inline tests/lax_scipy_test.py:559]: delete?

File: tests/lax_scipy_test.py [hunk 3]
```diff
@@ -576,12 +576,8 @@ def test_spectral_dac_eigh(self, linear_size, seed, dtype):
       'linear_size': linear_size, 'seed': seed, 'dtype': dtype}
     for linear_size in linear_sizes
     for seed in seeds
-    for dtype in jtu.dtypes.floating))
+    for dtype in [jnp.float32, ]))
```
Comment (rmlarsen) [inline tests/lax_scipy_test.py:579]: revert?

File: tests/lax_scipy_test.py [hunk 4]
```diff
@@ -576,12 +576,8 @@ def test_spectral_dac_eigh(self, linear_size, seed, dtype):
...
   def test_spectral_dac_svd(self, linear_size, seed, dtype):
```
Comment (rmlarsen) [inline tests/lax_scipy_test.py:580]: let's use "bisection" instead of "dac" (which is hardcoded to translate to "digital to analog converter" in my brain ;-) )

File: tests/lax_scipy_test.py [hunk 5]
```diff
@@ -59,7 +59,7 @@
 methods = ["qdwh", "svd"]
 seeds = [1, 10]
 
-linear_sizes = [16, 128, 256]
+linear_sizes = [8,] #128, 256]
```
Comment (rmlarsen) [inline tests/lax_scipy_test.py:62]: revert? We should test for larger sizes.

File: jax/_src/scipy/eigh.py [hunk 6]
```diff
@@ -190,6 +238,24 @@ def _eigh_work(
...
+    raise TypeError(f"Input H of shape {H.shape} must be square.")
+  if symmetrize:
+    H = 0.5 * (H + H.conj().T)
+
+  if ncols <= termination_size:
+    return jnp.linalg.eigh(H)
+
+  evals, evecs = _eigh_work(H, None, precision, termination_size)
+  sort_idxs = jnp.argsort(evals)
+  evals = evals[sort_idxs]
```
Comment (rmlarsen) [inline jax/_src/scipy/eigh.py:254]: out of curiosity: Does JAX have a fast O(n) in-place permute implementation?

File: jax/_src/scipy/eigh.py [hunk 7]
```diff
@@ -30,6 +56,62 @@ def _similarity_transform(
   return jnp.dot(out, matrix_out, precision=precision)
 
 
+def _canonically_purify_initial_guess(H, rank):
+  lambda_min, lambda_max = _gershgorin(H)
+  N = H.shape[0]
+  mu = jnp.trace(H) / N
+  theta = rank / N
+  beta = theta / (lambda_max - mu)
```
Comment (rmlarsen) [inline jax/_src/scipy/eigh.py:64]: We should probably guard against division by zero here. This would blow up for H=eye(n), I think.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by hawkinsp** [comment]
Closing. We've had a `jit`-table `eigh` for some time by using a bucketing and padding strategy.
